torch/_functorch/_aot_autograd/utils.py

```diff
@@ -234,59 +237,116 @@ def maybe_to_fresh_input(idx, t, meta):
+                torch.ops.prims._sink_tokens.default,
+                (output_token_nodes,),
+            )
+            node.args = (other_output_args,)
+
+    def do(module):
+        input_nodes = []
+        input_token_nodes = []
+        with_effect_nodes = []
+        output_token_nodes = []
+        other_output_nodes = []
+        for i, node in enumerate(module.graph.nodes):
+            if node.op == "placeholder":
+                input_nodes.append(node)
+            elif node.op == "call_function" and node.target.__name__ == "with_effects":
+                with_effect_nodes.append(node)
+                if node.args[0] in input_nodes:
+                    input_token_nodes.append(node.args[0])
+                    rewrite_with_effects_input_token(module, node)
+            elif node.op == "output":
+                outs = node.args[0]
+                for out in outs:
+                    if (
+                        isinstance(out, torch.fx.node.Node)
+                        and out.op == "call_function"
+                        and out.target == operator.getitem
+                        and out.args[1] == 0
+                        and out.args[0] in with_effect_nodes
+                    ):
+                        output_token_nodes.append(out)
+                    else:
+                        other_output_nodes.append(out)
+
+                rewrite_output(module, node, output_token_nodes, other_output_nodes)
+
+        for input_token_node in input_token_nodes:
+            module.graph.erase_node(input_token_node)
+
+        module.recompile()
+
+    if num_tokens > 0:
+        if aot_config.enable_log:
+            from torch._dynamo.utils import lazy_format_graph_code
+
+            aot_graphs_effects_log.info(
+                "%s",
+                lazy_format_graph_code(
+                    "Forward graph before unlifting tokens",
+                    fw_module,
+                    aot_config.aot_id,
+                    include_stride=True,
+                    include_device=True,
+                    colored=True,
+                ),
+            )
+        do(fw_module)
+
+    if bw_module is not None and num_bw_out_tokens > 0:
+        if aot_config.enable_log:
+            from torch._dynamo.utils import lazy_format_graph_code
+
+            aot_graphs_effects_log.info(
+                "%s",
+                lazy_format_graph_code(
+                    "Backward graph before unlifting tokens",
+                    bw_module,
+                    aot_config.aot_id,
+                    include_stride=True,
+                    include_device=True,
+                    colored=True,
+                ),
+            )
+        do(bw_module)
 
-    fw_module.recompile()
+    # No need to update fw_metadata.num_forward_returns and fw_metadata.num_forward as num_tokens are not part of it.
+    # As CompiledFunction.forward runs function wrapped in EffectTokensWrapper, that will toss out output tokens.
```

Comment:
This comment is confusing to me because in particular, in the path where we call `unlift_tokens` (when `config.unlift_effect_tokens` is set), the EffectTokensWrapper doesn't actually have to do anything at runtime since there are no extra token input/output runtime arguments